clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberState.java

```diff
@@ -17,57 +17,111 @@
+     * needed, invoke the user callbacks for onPartitionsAssigned or onPartitionsRevoked, and make
+     * the new assignment effective). Note that while in this state the member may be trying to
+     * resolve metadata for the target assignment, or triggering commits/callbacks if topic names
+     * already resolved.
      */
-    // TODO: determine if separate state will be needed for assign/revoke (not for now)
     RECONCILING,
 
     /**
-     * Member is active in a group (heartbeating) and has processed all assignments received.
+     * Member has completed reconciling an assignment received, and stays in this state only until
+     * the next heartbeat request is sent out to acknowledge the assignment to the server. This
+     * state indicates that the next heartbeat request must be sent without waiting for the
+     * heartbeat interval to expire. Note that once the ack is sent, the member could go back to
+     * {@link #RECONCILING} if it still has assignment waiting to be reconciled (assignments
+     * waiting for metadata, assignments for which metadata was resolved, or new assignments
+     * received from the broker)
+     */
+    ACKNOWLEDGING,
+
+    /**
+     * Member is active in a group and has processed all assignments received. While in this
+     * state, the member will send heartbeats on the interval.
      */
     STABLE,
 
     /**
-     * Member transitions to this state when it receives a
-     * {@link org.apache.kafka.common.protocol.Errors#UNKNOWN_MEMBER_ID} or
-     * {@link org.apache.kafka.common.protocol.Errors#FENCED_MEMBER_EPOCH} error from the
-     * broker. This is a recoverable state, where the member
-     * gives up its partitions by invoking the user callbacks for onPartitionsLost, and then
-     * transitions to {@link #UNJOINED} to rejoin the group as a new member.
+     * Member transitions to this state when it receives a {@link Errors#UNKNOWN_MEMBER_ID} or
+     * {@link Errors#FENCED_MEMBER_EPOCH} error from the broker, indicating that it has been
+     * left out of the group. While in this state, the member will stop sending heartbeats, it
+     * will give up its partitions by invoking the user callbacks for onPartitionsLost, and then
+     * transition to {@link #JOINING} to re-join the group as a new member.
      */
     FENCED,
 
     /**
-     * The member failed with an unrecoverable error
+     * The member transitions to this state after a call to unsubscribe. While in this state, the
+     * member will stop sending heartbeats, will commit offsets if needed and release its
+     * assignment (calling user's callback for partitions revoked or lost). When all these
+     * actions complete, the member will transition out of this state into {@link #LEAVING} to
+     * effectively leave the group.
+     */
+    PREPARE_LEAVING,
+
+    /**
+     * Member has committed offsets and releases its assignment, so it stays in this state until
```

Comment:
This is just a transient state for the member to send the leave group heartbeat with epoch -1/-2 (dynamic/static membership) right?

Reply:
Yes, similar to the ACKNOWLEDGING in the sense that they are just a way to indicate that a heartbeat must be sent without waiting for the interval, and as soon as the request is sent the member transitions out of the state.